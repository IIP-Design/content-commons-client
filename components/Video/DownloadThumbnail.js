import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import DownloadItem from './DownloadItem';
import { maybeGetUrlToProdS3 } from 'lib/utils';

const DownloadThumbnail = ( { instructions, units } ) => {
  const renderFormItem = ( unit, i ) => (
    <DownloadItem
      download={ `${unit.language.display_name}_thumbnail` }
      header={ `Download ${unit.language.display_name} Thumbnail` }
      hover={ `Download ${unit.language.display_name} Thumbnail` }
      key={ i }
      url={ maybeGetUrlToProdS3( unit.thumbnail ) }
    />
  );

  const renderFormItems = () => {
    const thumbnail = units
      .filter( unit => unit.thumbnail )
      .map( ( unit, i ) => renderFormItem( unit, i ) );

    return thumbnail.length ? thumbnail : 'There are no thumbnails available for download at this time';
  };

  return (
    <Fragment>
      <p className="form-group_instructions">{ instructions }</p>
      { units && renderFormItems( units ) }
    </Fragment>
  );
};

DownloadThumbnail.propTypes = {
  units: PropTypes.array,
  instructions: PropTypes.string,
};

export default DownloadThumbnail;
