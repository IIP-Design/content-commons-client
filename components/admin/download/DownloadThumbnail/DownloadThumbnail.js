import React, { Fragment } from 'react';
import { Item } from 'semantic-ui-react';
import { object, string } from 'prop-types';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import downloadIcon from 'static/icons/icon_download.svg';

const DownloadThumbnail = ( { instructions, data } ) => {
  const { error, loading, project } = data;

  if ( loading ) return 'Loading the project...';
  if ( error ) return `Error! ${error.message}`;

  if ( !project || !Object.keys( project ).length ) return null;

  const { units } = project;

  const renderFormItem = unit => {
    const { id, url, language: { displayName } } = unit;
    return (
      <Item.Group key={ `fs_${id}` } className="download-item">
        <Item as="a" href={ url } download={ `${displayName}_thumbnail` }>
          <Item.Image size="mini" src={ downloadIcon } className="download-icon" />
          <Item.Content>
            <Item.Header className="download-header">
              { `Download ${displayName} Thumbnail` }
            </Item.Header>
            <span className="item_hover">
              { `Download ${displayName} Thumbnail` }
            </span>
          </Item.Content>
        </Item>
      </Item.Group>
    );
  };

  const renderFormItems = () => {
    const t = units
      .filter( unit => unit && unit.url )
      .map( unit => renderFormItem( unit ) );
    return t.length ? t : 'There are no thumbnails available for download at this time';
  };

  return (
    <Fragment>
      <p className="form-group_instructions">{ instructions }</p>
      { units && renderFormItems( units ) }
    </Fragment>
  );
};

DownloadThumbnail.propTypes = {
  data: object,
  instructions: string
};

const VIDEO_PROJECT_PREVIEW_THUMBNAILS_QUERY = gql`
  query VideoProjectPreviewThumbnails($id: ID!) {
    project: videoProject(id: $id) {
      units: supportFiles(
        where: {
          OR: [
            { filetype: "jpg" },
            { filetype: "png" }
          ]
        },
        orderBy: filename_ASC
      ) {
        id
        url
        language {
          displayName
        }
      }
    }
  }
`;

export default graphql( VIDEO_PROJECT_PREVIEW_THUMBNAILS_QUERY, {
  options: props => ( {
    variables: {
      id: props.id
    },
  } )
} )( DownloadThumbnail );

export { VIDEO_PROJECT_PREVIEW_THUMBNAILS_QUERY };
