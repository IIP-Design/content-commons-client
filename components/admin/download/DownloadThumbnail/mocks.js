import { VIDEO_PROJECT_PREVIEW_THUMBNAILS_QUERY } from './DownloadThumbnail';

export const props = {
  id: '123',
  instructions: 'Download Thumbnail(s)',
  isPreview: false,
};

export const mocks = [
  {
    request: {
      query: VIDEO_PROJECT_PREVIEW_THUMBNAILS_QUERY,
      variables: { id: props.id },
    },
    result: {
      data: {
        project: {
          id: props.id,
          thumbnails: [
            {
              id: 'th765',
              url: `2019/06/${props.id}/thumbnail-1.jpg`,
              language: {
                id: 'en23',
                displayName: 'English',
              },
            },
            {
              id: 'th865',
              url: `2019/06/${props.id}/thumbnail-2.jpg`,
              language: {
                id: 'en23',
                displayName: 'English',
              },
            },
          ],
        },
      },
    },
  },
];

export const errorMocks = [
  {
    ...mocks[0],
    result: {
      errors: [{ message: 'There was an error.' }],
    },
  },
];

export const nullProjectMocks = [
  {
    ...mocks[0],
    result: { data: { project: null } },
  },
];

export const noFilesMocks = [
  {
    ...mocks[0],
    result: {
      data: {
        project: {
          ...mocks[0].result.data.project,
          thumbnails: [],
        },
      },
    },
  },
];

export const emptyProjectMocks = [
  {
    ...mocks[0],
    result: { data: { project: {} } },
  },
];
