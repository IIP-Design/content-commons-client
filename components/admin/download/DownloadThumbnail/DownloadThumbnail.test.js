import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from '@apollo/react-testing';
import { Loader } from 'semantic-ui-react';

import DownloadThumbnail from './DownloadThumbnail';

import {
  emptyProjectMocks,
  errorMocks,
  mocks,
  noFilesMocks,
  nullProjectMocks,
  props,
} from './mocks';

jest.mock( 'lib/utils', () => ( {
  getS3Url: jest.fn( assetPath => `https://s3-url.com/${assetPath}` ),
  getApolloErrors: error => {
    let errs = [];
    const { graphQLErrors, networkError, otherError } = error;

    if ( graphQLErrors ) {
      errs = graphQLErrors.map( error => error.message );
    }
    if ( networkError ) {
      errs.push( networkError );
    }

    if ( otherError ) {
      errs.push( otherError );
    }

    return errs;
  },
} ) );

jest.mock( 'static/icons/icon_download.svg', () => 'downloadIconSVG' );

jest.mock( 'lib/hooks/useSignedUrl', () => jest.fn( () => ( { signedUrl: 'https://example.jpg' } ) ) );

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <DownloadThumbnail { ...props } />
  </MockedProvider>
);

const suppressActWarning = consoleError => {
  const actMsg = 'Warning: An update to %s inside a test was not wrapped in act';

  jest.spyOn( console, 'error' ).mockImplementation( ( ...args ) => {
    if ( !args[0].includes( actMsg ) ) {
      consoleError( ...args );
    }
  } );
};

describe( '<DownloadThumbnail />', () => {
  const consoleError = console.error;

  beforeAll( () => suppressActWarning( consoleError ) );
  afterAll( () => {
    console.error = consoleError;
  } );

  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const downloadThumb = wrapper.find( 'DownloadThumbnail' );
    const loader = (
      <Loader
        active
        inline="centered"
        style={ { marginBottom: '1em' } }
        content="Loading thumbnail(s)..."
      />
    );

    expect( downloadThumb.exists() ).toEqual( true );
    expect( downloadThumb.contains( loader ) ).toEqual( true );
    expect( toJSON( downloadThumb ) ).toMatchSnapshot();
  } );

  it( 'renders error message if error is thrown', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename={ false }>
        <DownloadThumbnail { ...props } />
      </MockedProvider>,
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const downloadThumb = wrapper.find( 'DownloadThumbnail' );
    const errorComponent = downloadThumb.find( 'ApolloError' );

    expect( errorComponent.exists() ).toEqual( true );
    expect( errorComponent.contains( 'There was an error.' ) )
      .toEqual( true );
  } );

  it( 'renders final state without crashing', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const downloadThumb = wrapper.find( 'DownloadThumbnail' );

    expect( toJSON( downloadThumb ) ).toMatchSnapshot();
  } );

  it( 'renders null if project is null', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ nullProjectMocks } addTypename={ false }>
        <DownloadThumbnail { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadThumb = wrapper.find( 'DownloadThumbnail' );

    expect( downloadThumb.html() ).toEqual( null );
  } );

  it( 'renders null if project is {}', async () => {
    // ignore console.warn about missing field `id` and `thumbnails`
    const consoleWarn = console.warn;

    console.warn = jest.fn();

    const wrapper = mount(
      <MockedProvider mocks={ emptyProjectMocks } addTypename={ false }>
        <DownloadThumbnail { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadThumb = wrapper.find( 'DownloadThumbnail' );

    expect( downloadThumb.html() ).toEqual( null );
    console.warn = consoleWarn;
  } );

  it( 'renders a "no thumbnails available message" if there are no thumbnail files', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ noFilesMocks } addTypename={ false }>
        <DownloadThumbnail { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadThumb = wrapper.find( 'DownloadThumbnail' );
    const items = downloadThumb.find( 'Item' );
    const { thumbnails } = noFilesMocks[0].result.data.project;
    const msg = 'There are no thumbnails available for download at this time';

    expect( downloadThumb.exists() ).toEqual( true );
    expect( items.length ).toEqual( thumbnails.length );
    expect( downloadThumb.contains( msg ) ).toEqual( true );
  } );

  it( 'renders <a> tags with the correct href and download attribute values', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const items = wrapper.find( 'DownloadItemContent' );
    const { thumbnails } = mocks[0].result.data.project;
    const s3Bucket = 'https://s3-url.com';

    expect( items.length ).toEqual( thumbnails.length );
    
    items.forEach( ( item, i ) => {
      const assetPath = thumbnails[i].url;
      expect( item.prop( 'srcUrl' ) ).toEqual( `${s3Bucket}/${assetPath}` );
      expect( item.find( '.download-item' ).prop( 'href' ) ).toEqual( 'https://example.jpg' );
      expect( item.find( '.download-item' ).prop( 'download' ) ).toEqual( true );
    } );
  } );

  it( 'renders preview text if isPreview is true', async () => {
    const newProps = { ...props, isPreview: true };
    const wrapper = mount(
      <MockedProvider mocks={ mocks } addTypename={ false }>
        <DownloadThumbnail { ...newProps } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadThumbnail = wrapper.find( 'DownloadThumbnail' );
    const previews = downloadThumbnail.find( '.preview-text' );
    const { thumbnails } = mocks[0].result.data.project;

    expect( previews.length ).toEqual( thumbnails.length );
    previews.forEach( preview => {
      expect( preview.exists() )
        .toEqual( downloadThumbnail.prop( 'isPreview' ) );
      expect( preview.name() ).toEqual( 'span' );
      expect( preview.text() )
        .toEqual( 'The link will be active after publishing.' );
    } );
  } );

  it( 'does not render preview text if !isPreview', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const downloadThumbnail = wrapper.find( 'DownloadThumbnail' );
    const previews = downloadThumbnail.find( '.preview-text' );

    expect( previews.exists() )
      .toEqual( downloadThumbnail.prop( 'isPreview' ) );
    expect( previews.length ).toEqual( 0 );
  } );
} );
