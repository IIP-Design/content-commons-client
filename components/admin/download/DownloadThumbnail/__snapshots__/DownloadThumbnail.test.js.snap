// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DownloadThumbnail /> renders final state without crashing 1`] = `
<DownloadThumbnail
  data={
    Object {
      "called": true,
      "error": undefined,
      "fetchMore": [Function],
      "loading": false,
      "networkStatus": 7,
      "previousData": undefined,
      "project": Object {
        "id": "123",
        "thumbnails": Array [
          Object {
            "id": "th765",
            "language": Object {
              "displayName": "English",
              "id": "en23",
            },
            "url": "2019/06/123/thumbnail-1.jpg",
          },
          Object {
            "id": "th865",
            "language": Object {
              "displayName": "English",
              "id": "en23",
            },
            "url": "2019/06/123/thumbnail-2.jpg",
          },
        ],
      },
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "123",
      },
    }
  }
  id="123"
  instructions="Download Thumbnail(s)"
  isPreview={false}
>
  <DownloadItemContent
    hoverText="Download English Thumbnail"
    isAdminPreview={false}
    key="fs_th765"
    srcUrl="https://s3-url.com/2019/06/123/thumbnail-1.jpg"
  >
    <a
      className="download-item"
      download={true}
      href="https://example.jpg"
      key="https://example.jpg"
      rel="noopener noreferrer"
    >
      <div
        className="item-icon"
      >
        <img
          alt="Download English Thumbnail"
          src="downloadIconSVG"
        />
      </div>
      <div
        className="item-content"
      >
        <p
          className="item-content__title"
        >
          <strong>
            Download English Thumbnail
          </strong>
        </p>
      </div>
      <span
        className="item-hover"
      >
        Download English Thumbnail
      </span>
    </a>
  </DownloadItemContent>
  <DownloadItemContent
    hoverText="Download English Thumbnail"
    isAdminPreview={false}
    key="fs_th865"
    srcUrl="https://s3-url.com/2019/06/123/thumbnail-2.jpg"
  >
    <a
      className="download-item"
      download={true}
      href="https://example.jpg"
      key="https://example.jpg"
      rel="noopener noreferrer"
    >
      <div
        className="item-icon"
      >
        <img
          alt="Download English Thumbnail"
          src="downloadIconSVG"
        />
      </div>
      <div
        className="item-content"
      >
        <p
          className="item-content__title"
        >
          <strong>
            Download English Thumbnail
          </strong>
        </p>
      </div>
      <span
        className="item-hover"
      >
        Download English Thumbnail
      </span>
    </a>
  </DownloadItemContent>
</DownloadThumbnail>
`;

exports[`<DownloadThumbnail /> renders initial loading state without crashing 1`] = `
<DownloadThumbnail
  data={
    Object {
      "called": true,
      "error": undefined,
      "fetchMore": [Function],
      "loading": true,
      "networkStatus": 1,
      "previousData": undefined,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "123",
      },
    }
  }
  id="123"
  instructions="Download Thumbnail(s)"
  isPreview={false}
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "minHeight": "200px",
      }
    }
  >
    <Loader
      active={true}
      content="Loading thumbnail(s)..."
      inline="centered"
      style={
        Object {
          "marginBottom": "1em",
        }
      }
    >
      <div
        className="ui active text centered inline loader"
        style={
          Object {
            "marginBottom": "1em",
          }
        }
      >
        Loading thumbnail(s)...
      </div>
    </Loader>
  </div>
</DownloadThumbnail>
`;
